torchvision/datasets/forksafeziplookup.py

```diff
@@ -0,0 +1,46 @@
+import zipfile
+import threading
+import os, io
+import struct
+
+
+_FH_FILENAME_LENGTH = 10
+_FH_EXTRA_FIELD_LENGTH = 11
+STRUCT_FILE_HEADER = "<4s2B4HL2L2H"
+SIZE_FILE_HEADER = struct.calcsize(STRUCT_FILE_HEADER)
+
+
+# thread-safe/multiprocessing-safe
+class ForkSafeZipLookup:
+    def __init__(self, filename):
+        self.root_zip_filename = filename
+        self.root_zip_lookup = {}
+        self.root_zip = {}
+
+        with zipfile.ZipFile(filename, "r") as root_zip:
+            for info in root_zip.infolist():
+                if info.is_dir():
+                    continue
+                if info.compress_type != zipfile.ZIP_STORED:
+                    raise ValueError("Only uncompressed ZIP file supported: " + info.filename)
+                if info.compress_size != info.file_size:
+                    raise ValueError("Must be the same when uncompressed")
+                self.root_zip_lookup[info.filename] = (info.header_offset, info.compress_size)
+
+
+    def __getitem__(self, path):
+        key = (os.getpid(), threading.get_ident())
+        if not key in self.root_zip:
+            self.root_zip[key] = open(self.root_zip_filename, "rb")
```

Comment:
So we leave the file open forever? Is that recommended? Can't we just open the file until we read `io.BytesIO`?

Reply:
I have to say this is a matter of taste. The overhead of opening/closing the same file everytime is probably low (even on NFS). I tried to avoid this overhead, but it does add complexity.

If we do not keep the file open, but re-open the zip file everytime in __getitem__, then actually we can do away with obtaining the thread's identity (because the file descriptor becomes a local variable and we can't get into race conditions). Not 100% sure yet we can go without it and still be thread-safe/fork-safe, but I think it is.

I can guess the answer: do you want me to check if we can go without the complexity of thread-identity checking by not keeping the file descriptor open?

Reply:
I think I'd prefer to avoid this thread-identity checking, and open the zip file at every `__getitem__`. It might be slightly slower than keeping the file open, but I think that's worth it

Reply:
Resolved by re-opening the file each time.